### PR TITLE
Add framework for CoMeT broker stress testing

### LINKED ISSUE #44: Test framework providing realistic number of dummy clients
In order to bring #29 back and make it work, we need to be able to recreate it's problems in a test.
For this we should have a test framework that provides a configurable number of dummy clients, sending requests somehow realistically until the comet broker in branch `rn/redislocks` crashes.

One option would be to make these clients load real data from `recv1:/data/comet` and then act like a kotekan [datasetManager](http://lwlab.dunlap.utoronto.ca/kotekan/dev/dev_dataset_states.html):

1. register each dataset-state ID found in the dump, using `/register-state`
   - comet will respond the request it received first demanding the actual state.
   - the client that receives that request should send the state to `/send-state`
2. register each dataset ID found in the dump, using `/register-dataset`

- all clients should send the same request at about the same time
- we might want to add some clients as well that make us of `/update-datasets` and `/request-state` to query info from comet.
- Selecting a few of the dump files, even if they are in sequence, might include datasets that reference to base dataset IDs that are not in the files selected. That would break the comet broker, but it's not what we want to test.
- we could add a mode to shuffle the order the states and datasets are sent

### PR DESCRIPTION
**Pull request by anjakefala**
Closes #44.

### CODE REVIEW

**Comment by anjakefala** [comment]
Note: @nritsche comment here, that is still not addressed: https://github.com/chime-experiment/comet/pull/58#discussion_r348138918

**Comment by nritsche** [comment]
This will be great, thanks!
Some ideas on how to make it work better:
- On my machine this uses 100% on one core to send less than 1 request per second, while comet doesn't have to do much. Maybe locust is really heavy, otherwise we could try to make the rest a bit faster:
  - Skip the functions copied from `comet.manager`. Sending a line from the dump files as a json in a request directly should work if you skip the ones with `line["state"] = None`
  - I think the register_start and register_config make it slow down a lot when spawning a large number of clients. They can be removed.
- To simulate what made `comet` lock up in the past, we need a single state registered by a lot of clients at the same time. This state should be large.
- The CI tests fail because it is based of an old version of comet that uses a part of `chimedb` that has been removed. Merge or rebase on `rn/redislocks` to fix.
- For `/update-dataset` to be more stressfull for comet, we need to build dependencies between datasets: create one and use it as a base dataset for the next one. Then request an update on the tree you created,

**Comment by anjakefala** [comment]
Making this CI-able:
* Generate representative test states
* Add to .travis.yml a console-only run of locust

**Comment by anjakefala** [comment]
- [X] Skip the functions copied from comet.manager. Sending a line from the dump files as a json in a request directly should work if you skip the ones with line["state"] = None
- [X] I think the register_start and register_config make it slow down a lot when spawning a large number of clients. They can be removed.
- [ ] To simulate what made comet lock up in the past, we need a single state registered by a lot of clients at the same time. This state should be large.
- [ ]The CI tests fail because it is based of an old version of comet that uses a part of chimedb that has been removed. Merge or rebase on rn/redislocks to fix.
- [X] For /update-dataset to be more stressfull for comet, we need to build dependencies between datasets: create one and use it as a base dataset for the next one. Then request an update on the tree you created,

- [ ] It seems to prefer creating more requests over answering a request to send a state in the response to /register-state. This leads to requested states piling up in comet and requests depending on these states timing out. While this was nice as a stress test and helped finding bugs, it is (now that comet can deal better with it) hard to make sense out of the constantly long response times and possibly also contributes to the test client swarm being slow. On the other hand it is probably hard to fix, because as far as I understand we have no control over that.
- [ ] Generate representative test states, instead of using the sample ones from CoMeT runs
- [ ] Add to .travis.yml a console-only run of locust

File: tests/locust_stress_testing.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    def setup(self):
+        assert os.path.isfile(CHIMEDBRC), CHIMEDBRC_MESSAGE
+        os.environ["CHIMEDB_TEST_RC"] = CHIMEDBRC
+
+        # Make sure that we don't write to the actual chime database
+        os.environ["CHIMEDB_TEST_ENABLE"] = "Yes, please."
+
+        self.broker = Popen(
+            ["comet", "--debug", "1", "-t", "2", "-p", PORT]
```
Comment (nritsche) [inline tests/locust_stress_testing.py]: Let's make this faster:
- if we don't need it for some reason: `--debug 0`
- `-w ($TRAVIS_NUMCORES - 1)` (?)

**Comment by nritsche** [comment]
Oh, looks like this is running out of memory:
```
[2019-11-21 18:45:22 +0000] [8630] [ERROR] Exception occurred while handling uri: 'http://localhost:8000/send-state'

Traceback (most recent call last):

  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/sanic/request.py", line 169, in load_json

    self.parsed_json = loads(self.body)

ValueError: Could not reserve memory block
```

File: .travis.yml [hunk 2]
```diff
@@ -49,7 +50,7 @@ before_script:
 script:
     # Run pycodestyle on all .py files in all subfolders
     - black --check .
-    - find . -type d -o ! -name versioneer.py ! -name test_*.py ! -name _version.py -name \*.py -exec pydocstyle --convention=numpy --add-ignore=D105,D202 {} +
+    - find . -type d -o ! -name versioneer.py ! -name locust_*.py ! -name test_*.py ! -name _version.py -name \*.py -exec pydocstyle --convention=numpy --add-ignore=D105,D202 {} +
```
Comment (nritsche) [inline .travis.yml]: Or document the tests...

Comment (anjakefala) [reply .travis.yml]: Added!

**Comment by anjakefala** [comment]
Ready for review

File: tests/locust_stress_testing.py [hunk 3]
```diff
@@ -140,7 +140,7 @@ class DummyManager(DummyClientLocust):
 
     host = HOST
     port = PORT
-    wait_time = lambda x: 3
+    wait_time = lambda x: 20
```
Comment (nritsche) [inline tests/locust_stress_testing.py]: Did this work? Doesn't this mean less stress? But on the other hand, will the problem that this tries to solve reappear if we add clients again?

Comment (anjakefala) [reply tests/locust_stress_testing.py]: It increases the wait time between new tasks, which I think does increase the space for clients to finish a full task before the broker is bombarded.

But yeah, it does mean less stress. I do not think there is a better way for us to address this. =( There is a tension between adding clients, increasing client activity, and making sure clients have space to send states upon request.

Comment (anjakefala) [reply tests/locust_stress_testing.py]: We can also make this an actual function, where a random wait time in a range is chosen. That might be a good middle ground.

**Pull request merged**